From: David Howells <dhowells@redhat.com>
Date: Tue, 23 Jul 2013 16:49:24 +0100
Subject: [PATCH] Fix __wait_on_atomic_t() to call the action func if the
 counter != 0
Git-commit: 42577ca8c3616baaafdd8f167b2e1fb959026081
Patch-mainline: v3.11
References: bnc#880344

Fix __wait_on_atomic_t() so that it calls the action func if the counter != 0
rather than if the counter is 0 so as to be analogous to __wait_on_bit().

Thanks to Yacine who found this by visual inspection.

This will affect FS-Cache in that it will could fail to sleep correctly when
trying to clean up after a netfs cookie is withdrawn.

Reported-by: Yacine Belkadi <yacine.belkadi.1@gmail.com>
Signed-off-by: David Howells <dhowells@redhat.com>
Reviewed-by: Jeff Layton <jlayton@redhat.com>
Cc: Milosz Tanski <milosz@adfin.com>
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
Acked-by: NeilBrown <neilb@suse.de>

---
 kernel/wait.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- linux-3.0-SLE11-SP3.orig/kernel/wait.c
+++ linux-3.0-SLE11-SP3/kernel/wait.c
@@ -332,7 +332,8 @@ int __wait_on_atomic_t(wait_queue_head_t
 		prepare_to_wait(wq, &q->wait, mode);
 		val = q->key.flags;
 		if (atomic_read(val) == 0)
-			ret = (*action)(val);
+			break;
+		ret = (*action)(val);
 	} while (!ret && atomic_read(val) != 0);
 	finish_wait(wq, &q->wait);
 	return ret;
